### PR TITLE
Added feature config groups

### PR DESCRIPTION
**Pull request by KeshavSoni2511**
Fix for issue #752

### CODE REVIEW

**Comment by Keshav-NEC** [comment]
Hi @fgalan , the tests are passing on my local setup but here the tests are failing. Kindly look into the issue or anything I can fix. Thanks

**Comment by mapedraza** [comment]
> IMG_4038 Hi @fgalan , the tests are passing on my local setup but here the tests are failing. Kindly look into the issue or anything I can fix. Thanks

Hi @Keshav-NEC , thank for your willingness to contribute!

The tests only fails on Node V18, which version are you using in your local enviroment?. Would be great if you can review it using the same version (V18). I suggest to use [`n` Node version manager](https://github.com/tj/n) to switch between different node versions 

Probably those errors are related with https://github.com/telefonicaid/iotagent-node-lib/pull/1333

**Comment by Keshav-NEC** [comment]
Hi @mapedraza , I have tested the setup on `Node V16`. I will test the setup on `Node V18`  and will share my findings.

**Comment by Keshav-NEC** [comment]
Hi @fgalan , @mapedraza I have tested the setup on ```Node V18``` and ```npm V9.5.1``` and test cases are passing and also I have used the link [`n` Node version manager](https://github.com/tj/n) shared by @mapedraza . Please find the logs of test cases of pull request #1375 and also I have attached the screenshot of passing test cases. Thanks.
Screenshot 2023-06-21 000549

File: test/unit/mongodb/mongodb-configGroup-registry-test.js [hunk 1]
```diff
@@ -0,0 +1,451 @@
...
+const should = require('should');
+const iotAgentConfig = {
+    logLevel: 'FATAL',
+    contextBroker: {
+        host: '192.168.1.1',
+        port: '1026'
+    },
+    server: {
+        name: 'testAgent',
+        port: 4041,
```
Comment (mapedraza) [inline test/unit/mongodb/mongodb-configGroup-registry-test.js:42]:
```suggestion
        port: 4041,
        host: 'localhost',
```

File: test/unit/ngsiv2/provisioning/device-provisioning-configGroup-api_test.js [hunk 2]
```diff
@@ -0,0 +1,1327 @@
...
+let contextBrokerMock;
+const iotAgentConfig = {
+    logLevel: 'FATAL',
+    contextBroker: {
+        host: '192.168.1.1',
+        port: '1026',
+        ngsiVersion: 'v2'
+    },
+    server: {
+        port: 4041,
```
Comment (mapedraza) [inline test/unit/ngsiv2/provisioning/device-provisioning-configGroup-api_test.js:43]:
```suggestion
        port: 4041,
        host: 'localhost',
```

File: test/unit/ngsiv2/provisioning/singleGroupConfigurationMode-test.js [hunk 3]
```diff
@@ -0,0 +1,320 @@
...
+const request = require('request');
+const iotAgentConfig = {
+    logLevel: 'FATAL',
+    contextBroker: {
+        host: '192.168.1.1',
+        port: '1026',
+        ngsiVersion: 'v2'
+    },
+    server: {
+        port: 4041,
```
Comment (mapedraza) [inline test/unit/ngsiv2/provisioning/singleGroupConfigurationMode-test.js:42]:
```suggestion
        port: 4041,
        host: 'localhost',
```

**Comment by mapedraza** [comment]
As mentioned in my previous comment, you did not addressed the changes made in https://github.com/telefonicaid/iotagent-node-lib/pull/1333 in your new tests.

Take as an example the ones I noted above and check the rest of the codebase (I suggested 3, the tests indicates 5, so 2 are remaining)

**Comment by Keshav-NEC** [comment]
> As mentioned in my previous comment, you did not addressed the changes made in #1333 in your new tests.
> 
> Take as an example the ones I noted above and check the rest of the codebase (I suggested 3, the tests indicates 5, so 2 are remaining)

Hi @mapedraza , Thanks for the help and I have done changes as suggested by you and I am sharing the test cases screenshot.
Screenshot 2023-06-26 214131
Screenshot 2023-06-26 215608

**Comment by Keshav-NEC** [comment]
Hi @mapedraza, If you got any chance to review this PR.

**Comment by mapedraza** [comment]
There are still tests failling. You need to solve them before merging this PR

File: lib/services/northBound/deviceGroupAdministrationServer.js [hunk 4]
```diff
@@ -288,6 +306,35 @@ function loadContextRoutes(router) {
...
+    );
+
+    router.delete(
+        '/iot/configGroups',
+        restUtils.checkRequestAttributes('headers', mandatoryHeaders),
+        restUtils.checkRequestAttributes('query', mandatoryParameters),
+        handleDeleteDeviceGroups
+    );
+}
+
```
Comment (mapedraza) [inline lib/services/northBound/deviceGroupAdministrationServer.js:337]: The route is uppercase aware? Did you try requesting to  `/iot/configgroups`?

File: doc/api.md [hunk 5]
```diff
@@ -1140,11 +1140,50 @@ Successful operations return `Content-Type` header with `application/json` value
...
+                {
+                    "name": "swithc_status",
+                    "type": "boolean"
+                }
+            ]
+        }
+    ]
+}
+```
+
```
Comment (mapedraza) [inline doc/api.md:1186]: This should replace the existing example, not adding a new one.

File: doc/api.md [hunk 6]
```diff
@@ -1198,11 +1237,38 @@ _**Request headers**_
...
+                    }
+                }
+            ],
+            "lazy": [{ "name": "status", "type": "Boolean" }]
+        }
+    ]
+}
+```
+OR
+
```
Comment (mapedraza) [inline doc/api.md:1271]: This should replace the existing example, not adding a new one

File: lib/services/groups/groupRegistryMemory.js [hunk 7]
```diff
@@ -87,6 +87,40 @@ function getConfigurationsByService(service) {
...
+            }
+        }
+    }
+
+    callback(null, {
+        count: filteredGroups.length,
+        configGroups: result
+    });
+}
+
```
Comment (mapedraza) [inline lib/services/groups/groupRegistryMemory.js:123]: This is duplicated, same as `function listGroups(service, limit, offset, callback) `. No needed to duplicate code

File: lib/services/groups/groupRegistryMemory.js [hunk 8]
```diff
@@ -152,6 +186,32 @@ function findSingleConfigurationMode(service, subservice, callback) {
...
+    if (result.length > 0) {
+        return callback(null, {
+            count: result.length,
+            configGroups: result
+        });
+    } else {
+        return callback(new errors.DeviceGroupNotFound(service, subservice));
+    }
+}
+
```
Comment (mapedraza) [inline lib/services/groups/groupRegistryMemory.js:214]: This is duplicated, same as `function find(service, subservice, callback)`.  No needed to duplicate code

File: lib/services/groups/groupRegistryMongoDB.js [hunk 9]
```diff
@@ -119,6 +119,45 @@ function createGroup(group, callback) {
...
+        error,
+        results
+    ) {
+        callback(error, {
+            count: results[1],
+            configGroups: results[0].map(toObjectFn)
+        });
+    });
+}
+
```
Comment (mapedraza) [inline lib/services/groups/groupRegistryMongoDB.js:160]: Duplicated code, same as `function listGroups(service, limit, offset, callback)`. It is not required.

File: lib/services/groups/groupRegistryMongoDB.js [hunk 10]
```diff
@@ -182,6 +221,36 @@ function getById(id, callback) {
...
```
Comment (mapedraza) [inline lib/services/groups/groupRegistryMongoDB.js:253]: Duplicated code, same as `function find(service, subservice, callback) `. No required

File: lib/services/groups/groupRegistryMongoDB.js [hunk 11]
```diff
@@ -302,8 +371,10 @@ function clear(callback) {
 
 exports.create = alarmsInt(constants.MONGO_ALARM + '_01', intoTrans(context, createGroup));
 exports.list = alarmsInt(constants.MONGO_ALARM + '_02', intoTrans(context, listGroups));
+exports.listConfigGroups = alarmsInt(constants.MONGO_ALARM + '_12', intoTrans(context, listConfigGroups));
 exports.init = alarmsInt(constants.MONGO_ALARM + '_03', intoTrans(context, init));
 exports.find = alarmsInt(constants.MONGO_ALARM + '_04', intoTrans(context, find));
+exports.findConfigGroups = alarmsInt(constants.MONGO_ALARM + '_13', intoTrans(context, findConfigGroups));
 exports.findType = alarmsInt(
```
Comment (mapedraza) [inline lib/services/groups/groupRegistryMongoDB.js:378]: Exporting duplicated functions, not required

File: lib/services/northBound/deviceGroupAdministrationServer.js [hunk 12]
```diff
@@ -118,11 +123,19 @@ function applyConfigurationMiddlewares(newConfiguration, callback) {
...
+            req.body.configGroups[i].service = req.headers['fiware-service'];
+            req.body.configGroups[i].subservice = req.headers['fiware-servicepath'];
+        }
+    } else {    
+        for (let i = 0; i < req.body.services.length; i++) {
+            req.body.services[i] = applyMap(apiToInternal, req.body.services[i]);
+            req.body.services[i].service = req.headers['fiware-service'];
+            req.body.services[i].subservice = req.headers['fiware-servicepath'];
+        }
+   }
```
Comment (mapedraza) [inline lib/services/northBound/deviceGroupAdministrationServer.js:138]: The logic of different routes ( `/iot/services` and `/iot/configGroups` should be the same)

**Review by mapedraza — CHANGES_REQUESTED** [review]
As general comment, all the logic is duplicated for `/iot/services` and `/iot/configGroups`. It is quite more easy. The logic should be the same, it is enough routing `/iot/configGroups` to same methods than `/iot/services`

**Comment by Keshav-NEC** [comment]
Hi @mapedraza, As in this comment, https://github.com/telefonicaid/iotagent-node-lib/issues/752#issue-403146821 it is suggested to keep the backward compatibility intact.
I have created the separate functions for `/iot/services` and `/iot/configGroups` to return the result in **services:[]** and **configGroups:[]** objects respectively. 
`listGroups()` will output the result into **services: []** object [here ](https://github.com/telefonicaid/iotagent-node-lib/pull/1375/files#diff-99f2913683b9206e83df4dd00ac8adc87ff81d363c4a44b0ee913ffa32a238eeR154) but `listConfigGroups() `will output the result into **configGroups:[]** object [here](https://github.com/telefonicaid/iotagent-node-lib/pull/1375/files#diff-99f2913683b9206e83df4dd00ac8adc87ff81d363c4a44b0ee913ffa32a238eeR120).

 

As per @mapedraza, I need to delete the `listConfigGroups()` and other functions which are similar to previous one. And handle both APIs `/iot/services` and `/iot/configGroups` in the same function by using if/else or some other logic. Could you please confirm my understanding?

**Comment by mapedraza** [comment]
@Keshav-NEC as you stated, the idea is to not break previous methods and paths. My comment comes because of the duplicated logic you added:

```js
function listGroups(service, limit, offset, callback) {
    const result = [];
    let skipped = 0;
    const filteredGroups = getConfigurationsByService(service);
    for (const i in filteredGroups) {
        if (registeredGroups.hasOwnProperty(filteredGroups[i])) {
            if (offset && skipped < parseInt(offset, 10)) {
                skipped++;
            } else {
                result.push(registeredGroups[filteredGroups[i]]);
            }
            if (limit && result.length === parseInt(limit, 10)) {
                break;
            }
        }
    }
    callback(null, {
        count: filteredGroups.length,
        services: result
    });
}
```

```js
function listConfigGroups(service, limit, offset, callback) {
    const result = [];
    let skipped = 0;
    const filteredGroups = getConfigurationsByService(service);

    for (const i in filteredGroups) {
        if (registeredGroups.hasOwnProperty(filteredGroups[i])) {
            if (offset && skipped < parseInt(offset, 10)) {
                skipped++;
            } else {
                result.push(registeredGroups[filteredGroups[i]]);
            }

            if (limit && result.length === parseInt(limit, 10)) {
                break;
            }
        }
    }

    callback(null, {
        count: filteredGroups.length,
        configGroups: result
    });
}
```

Both codeblocks mentioned above differs just in one line, the name of the variable passed to the callback. Probably there are better ways to implement it without duplicating code, that supposes worst maintainability (duplicating logics and paths). Take the solution you feel confident with it (for example, a function wrapper could be an alternative) but try not to duplicate the code.

**Comment by AlvaroVega** [comment]
What is the difference between this PR and these others: https://github.com/telefonicaid/iotagent-node-lib/pull/1198 and https://github.com/telefonicaid/iotagent-node-lib/pull/954.
I guess two of there PRs should be closed.

**Comment by Keshav-NEC** [comment]
Hi @mapedraza , I have updated the PR as per the review comments received. Please review, Thanks.

File: doc/api.md (hunk 5 above)
Comment (Keshav-NEC) [reply doc/api.md]: Fixed in https://github.com/telefonicaid/iotagent-node-lib/pull/1375/commits/11787050de050581b6b76221f20d9c2c671c3b11

File: doc/api.md (hunk 6 above)
Comment (Keshav-NEC) [reply doc/api.md]: Fixed in https://github.com/telefonicaid/iotagent-node-lib/pull/1375/commits/11787050de050581b6b76221f20d9c2c671c3b11

File: lib/services/groups/groupRegistryMemory.js (hunk 7 above)
Comment (Keshav-NEC) [reply lib/services/groups/groupRegistryMemory.js]: Fixed in https://github.com/telefonicaid/iotagent-node-lib/pull/1375/commits/11787050de050581b6b76221f20d9c2c671c3b11

File: lib/services/groups/groupRegistryMemory.js (hunk 8 above)
Comment (Keshav-NEC) [reply lib/services/groups/groupRegistryMemory.js]: Fixed in https://github.com/telefonicaid/iotagent-node-lib/pull/1375/commits/11787050de050581b6b76221f20d9c2c671c3b11

File: lib/services/groups/groupRegistryMongoDB.js (hunk 9 above)
Comment (Keshav-NEC) [reply lib/services/groups/groupRegistryMongoDB.js]: Fixed in https://github.com/telefonicaid/iotagent-node-lib/pull/1375/commits/11787050de050581b6b76221f20d9c2c671c3b11

File: lib/services/groups/groupRegistryMongoDB.js (hunk 10 above)
Comment (Keshav-NEC) [reply lib/services/groups/groupRegistryMongoDB.js]: Fixed in https://github.com/telefonicaid/iotagent-node-lib/pull/1375/commits/11787050de050581b6b76221f20d9c2c671c3b11

File: lib/services/groups/groupRegistryMongoDB.js (hunk 11 above)
Comment (Keshav-NEC) [reply lib/services/groups/groupRegistryMongoDB.js]: Fixed in https://github.com/telefonicaid/iotagent-node-lib/pull/1375/commits/11787050de050581b6b76221f20d9c2c671c3b11

File: lib/services/northBound/deviceGroupAdministrationServer.js (hunk 12 above)
Comment (Keshav-NEC) [reply lib/services/northBound/deviceGroupAdministrationServer.js]: Fixed in https://github.com/telefonicaid/iotagent-node-lib/pull/1375/commits/11787050de050581b6b76221f20d9c2c671c3b11

**Comment by KeshavSoni2511** [comment]
Hi @mapedraza, If you got any chance to review this PR.

**Comment by KeshavSoni2511** [comment]
Hi @fgalan, if you got any chance to review this PR

**Review by mapedraza — CHANGES_REQUESTED** [review]
Hi [@KeshavSoni2511](https://github.com/KeshavSoni2511), than you so much for your amazing job. Now it is more aligned. Just some minor improvements points:

- [ ] 1): I can't see tests implementing cross functionality. I mean, creating groups using services and then retrieving it through configGroup endpoint. It should be aded in both ways, I mean, creating as services and configGroup and retrieving them with the opposite.
- [ ] 2) In order to configure the keyword used by the new endpoint, it would be great if instead of having a magic word, it is moved to the constans file (lib/constants.js), replacing each time in the code appears 'configgroups'. It should be configured for the API route and for the term used in the json response.

**Comment by mapedraza** [comment]
Hi @KeshavSoni2511, could you address what I noted in my previous comment?

> Hi [@KeshavSoni2511](https://github.com/KeshavSoni2511), than you so much for your amazing job. Now it is more aligned. Just some minor improvements points:
> 
> * [ ]  1): I can't see tests implementing cross functionality. I mean, creating groups using services and then retrieving it through configGroup endpoint. It should be aded in both ways, I mean, creating as services and configGroup and retrieving them with the opposite.
> * [ ]  2) In order to configure the keyword used by the new endpoint, it would be great if instead of having a magic word, it is moved to the constans file (lib/constants.js), replacing each time in the code appears 'configgroups'. It should be configured for the API route and for the term used in the json response.

**Comment by KeshavSoni2511** [comment]
> Hi @KeshavSoni2511, could you address what I noted in my previous comment?. It would be great if you complete everything so we can merge it
> 
> Thanks in advance!

Hi @mapedraza, Hope you are doing well. FYI, due to sudden layoff, I am not working with NEC anymore. I think anyone else from NEC can do requested changes.

**Comment by mapedraza** [comment]
> > Hi @KeshavSoni2511, could you address what I noted in my previous comment?. It would be great if you complete everything so we can merge it
> > Thanks in advance!
> 
> Hi @mapedraza, Hope you are doing well. FYI, due to sudden layoff, I am not working with NEC anymore. I think anyone else from NEC can do requested changes.

Really sad to hear that. I will merge your PR into a prelanding branch, so the work you accomplished is not going to be be thrown. 

Thank for your help

**Pull request merged**